### PR TITLE
[INFRA] GitHub CI

### PR DESCRIPTION
**Pull request by eseiler**
We can set travis to not build on pushed branches and pull requests. 
I would then maybe use travis to run the CI as a nightly on multiple CPU architectures.

Note:
The build-folder of the documentation (containing doc_dev and doc_usr) is provided as an artefact after the job ran. You can download it from the `checks` tab and then click `artifacts` on the top right or `SeqAn3 CI` on the left and find it there.

The artefact is only available after the **whole** (all test suites / compilers - not just the documentation job) GitHub actions run is complete - this is a limitation on GitHub's side.

The artifact will always be zipped (also a limitation on GitHub's side). That means the documentation will be provided as `documentation.tar.gz.zip`.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        with:
+          access_token: ${{ github.token }}
+  build:
+    needs: cancel
+    name: ${{ matrix.config.name }}
+    runs-on: ubuntu-20.04
+    strategy:
+      fail-fast: false
+      matrix:
+        config:
```
[the remark this answers is not in the document]
Comment (rrahn) [reply .github/workflows/ci.yml:26]: Here we would add the default for every job configuration in the matrix. At least this is my best bet after reading the api documentation.

**Review by rrahn — COMMENTED** [review]
I have some suggestions. Otherwise this looks very nice. Thank you for your efforts on this.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+      - name: Install GCC 9
+        if: matrix.config.cxx == 'g++-9'
+        shell: bash
+        run: sudo apt-get install --yes g++-9
+
+      - name: Install GCC 10
+        if: matrix.config.cxx == 'g++-10'
+        shell: bash
+        run: sudo apt-get install --yes g++-10
```
[the remark this answers is not in the document]
Comment (eseiler) [reply .github/workflows/ci.yml]: Sure, the `${{ }}` stuff is yaml and evaluated in the very beginning of the workflow

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+          cd seqan3-build
+          cmake ../seqan3/test/${BUILD} -DCMAKE_BUILD_TYPE=${BUILD_TYPE} -DCMAKE_CXX_FLAGS="${CXXFLAGS}"
+          if [[ "${BUILD}" =~ ^(unit|header|snippet|coverage)$ ]]; then
+            make gtest_project
+          fi
+          if [[ "${BUILD}" =~ ^(performance)$ ]]; then
+            make gbenchmark_project
+          fi
+
+      - name: Build and run tests
```
[the remark this answers is not in the document]
Comment (eseiler) [reply .github/workflows/ci.yml]: I split it, if one step fails the job returns as failed, so we don't need the check.

All steps have an implicit `if: ${{ success }}`, i.e. only run if previous step succeeded

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+              ctest . --output-on-failure]; else
+              ctest . -j2 --output-on-failure
+            fi
+          fi
+          ccache -s || true
+
+      - name: Submit coverage build
+        if: matrix.config.build == 'coverage'
+        shell: bash
+        run: bash <(curl -s https://codecov.io/bash) -f ./seqan3-build/seqan3_coverage -R ./seqan3 || echo 'Codecov failed to upload'
```
[the remark this answers is not in the document]
Comment (eseiler) [reply .github/workflows/ci.yml:215]: Yes, unless the echo fails :)

You could also do another step with `if: ${{ failure() }}` and do something about the error, but I don't think that's necessary

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+      - name: Tar documentation
+        if: matrix.config.build == 'documentation'
+        run: tar -zcf documentation.tar.gz seqan3-build
```
[the remark this answers is not in the document]
Comment (eseiler) [reply .github/workflows/ci.yml:220]: I allowed it to fail without reporting the job as failure

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+
+      - name: Upload documentation
```
[the remark this answers is not in the document]
Comment (eseiler) [reply .github/workflows/ci.yml:222]: I allowed it to fail without reporting the job as failure

**Comment by eseiler** [comment]
I think I incorporated all your suggestions.

As it turns out, it's really hard to set the default and then overwrite it.
The syntax

```
  matrix:
    require_something: [true]
    config:
      - { require_something: false
        }
```

has the problem, that there are `matrix.require_something` and `matrix.config.require_something`. So you are changing the value you don't want to change, but adding a new one.

But I found a way to declare the matrix with `include`, I think this looks cleaner.

If you would do the `require_something: [true]` and then set it somewhere to `false` it would be at the right scope.
But somehow, if I set it to false for documentation, only the header test and the documentation is built (i.e., the last in the include list where require_something=true).

I don't know if it's supposed to be like this but it kinda looks like the include is aggregated in a wrong way?
It probably should not be joined via `require_something`, but the individual config values should also be compared...

**Comment by eseiler** [comment]
@rrahn polite ping; I'd like to get this in this week so that it is merged into master beginning of next week

**Review by smehringer — APPROVED** [review]

**Pull request merged**